core/services/ocrcommon/data_source.go

```diff
@@ -176,19 +194,94 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
 
 // Observe without saving to DB
 func (ds *inMemoryDataSource) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
-	_, finalResult, err := ds.executeRun(ctx, ObservationTimestamp{
+	_, trrs, err := ds.executeRun(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	finalResult := trrs.FinalResult(ds.lggr)
+	setEATelemetry(ds, trrs, finalResult, ObservationTimestamp{
 		Round:        timestamp.Round,
 		Epoch:        timestamp.Epoch,
 		ConfigDigest: timestamp.ConfigDigest.Hex(),
 	})
-	if err != nil {
-		return nil, err
-	}
+
 	return ds.parse(finalResult)
 }
 
+// inMemoryDataSourceCache is a time based cache wrapper for inMemoryDataSource.
+// If cache update is overdue Observe defaults to standard inMemoryDataSource behaviour.
+type inMemoryDataSourceCache struct {
+	cacheExpiration time.Duration
+	mu              *sync.RWMutex
+	lastUpdate      atomic.Pointer[time.Time]
+	latestTrrs      pipeline.TaskRunResults
+	latestResult    pipeline.FinalResult
+	*inMemoryDataSource
+}
+
+// isUpdateOverdue atomically checks if update is overdue.
+func (ds *inMemoryDataSourceCache) isUpdateOverdue() bool {
+	timeSinceUpdate := time.Since(*ds.lastUpdate.Load())
+	return timeSinceUpdate >= ds.cacheExpiration
+}
+
+// updater periodically updates data source cache.
+func (ds *inMemoryDataSourceCache) updater() {
+	ticker := time.NewTicker(ds.cacheExpiration)
+	for range ticker.C {
+		// arbitrary timeout
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute*1)
+		if err := ds.updateCache(ctx); err != nil {
+			ds.lggr.Warnw("failed to update cache", "err", err)
+		}
+		cancel()
+	}
+}
+
+func (ds *inMemoryDataSourceCache) updateCache(ctx context.Context) error {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+	if ds.isUpdateOverdue() {
+		_, trrs, err := ds.executeRun(ctx)
+		if err != nil {
+			return errors.Wrapf(err, "error executing run for spec ID %v", ds.spec.ID)
+		}
+		lastUpdate := time.Now()
+
+		ds.latestResult = trrs.FinalResult(ds.lggr)
+		ds.latestTrrs = trrs
+		ds.lastUpdate.Store(&lastUpdate)
+	}
+	return nil
+}
+
+func (ds *inMemoryDataSourceCache) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
+	// if cache update is overdue revert to standard inMemoryDataSource Observe behaviour
+	result := make(chan error)
+	go func() { result <- ds.updateCache(ctx) }()
```

Comment:
whats going on here? it seems like you avhe 2 routines updating the cache? one here and one via `updater`?

i'd update the cache in one routine and just read the value of the cache here

Reply:
The idea was to minimize outdated values, there are two possibilities here:

1. `updater` successfully updated cache and `updateCache` just immediately returns from this go routine
2. `updater` failed to update cache and this new go routine that is spawned in `Observe` updates cache before returning values

Reply:
I can do the same thing but hide it behind cache get